connmgr/conngating.go

```diff
@@ -0,0 +1,71 @@
+package connmgr
+
+import (
+	"github.com/libp2p/go-libp2p-core/control"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/transport"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// ConnectionGater can be implemented by a type that supports active
+// inbound or outbound connection gating.
+//
+// A ConnectionGater will be consulted during different states in the life-cycle of a connection and
+// the specific gating function that will be called depends on the life-cycle state of the
+// connection. Hence, it is important to implement this interface keeping in mind the specific life-cycle state
+// at which you'd like to gate/block the connection.
+//
+// `InterceptDial` and `InterceptPeerDial` are called when we try an outbound dial.
+//
+// `InterceptAccept` is called after we've accepted an inbound connection from a socket but before we
+// begin upgrading it.
+//
+// `InterceptSecured` is called for inbound and outbound connections after we've negotiated
+// the security protocol to use for the connection.
+//
+// `InterceptUpgraded` is called for inbound and outbound connections after we've finished upgrading
+// a connection to have both security and stream multiplexing.
+//
+// This feature can be used to implement *strict* connection management
+// behaviours, such as hard limiting of connections once a max count has been
+// reached.
+//
+// If you'd like to send a disconnect control message to the remote peer for a gated inbound connection,
+// ONLY `InterceptUpgraded` should reject the connection (with an appropriate disconnect reason).
+// All other methods should allow the connection as we can ONLY open control streams
+// for upgraded connections.
+// Note: There's no point in sending disconnect control messages for outbound connections, so we might
+// as well close them as early in the cycle as possible.
+type ConnectionGater interface {
+	// InterceptDial tests whether we're permitted to dial the specified multiaddr.
+	// Insofar filter.Filters is concerned, this would map to its AddrBlock method,
+	// with the inverse condition.
+	// This is to be called by the network/swarm when dialling.
+	InterceptDial(ma.Multiaddr) (allow bool)
```

Comment:
This should pass in the target peer ID as well so we can do things like forbid/allow specific transports/ips to specific peers.

Reply:
Take a look at `InterceptPeerDial `.

Reply:
That will let me intercept the entire dial to a peer, but it doesn't let me decide "for this peer, should I allow this transport/addr/etc." For example, I may want to relay all outbound connections through a single relay in that case, I'd want to:

1. Allow relay multiaddrs through the target relay.
2. Allow _all_ multiaddrs for the target relay.

Reply:
@Stebalien I agree with you. Having both the peerId and the addrs here will be helpful here. I am making the change.